README.md
# `game-of-life` Package for GNU-Octave

[![DOI](https://zenodo.org/badge/509427410.svg)](https://zenodo.org/badge/latestdoi/509427410)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/aumpierre-unb/Game-of-Life-for-GNU-Octave)

<!-- ![Illustrative graphical output](https://github.com/aumpierre-unb/Game-of-Life-for-GNU-Octave/blob/main/pics/D2fRe.png "Example of graphical output") -->

![Illustrative graphical output](doc/ezgif.com-gif-maker.gif)

## Installing and Loading `game-of-life`

`game-of-life` is no registered at GNU Octave's repository. 

In order to install `game-of-life`, download the latest version compressed source code at [Releases](https://github.com/aumpierre-unb/Game-of-Life-for-GNU-Octave/releases), set the current directory to the download folder and call the following sequence from your console:

```dotnetcli
# e.g. this call will install version 0.1.2
# for latter versions, adapt the call accordingly
pkg install v0.1.2.tar.gz
pkg load game-of-life
```

## Citation of `game-of-life`

You can cite all versions (both released and pre-released), by using
[DOI 105281/zenodo.7395871](https://doi.org/10.5281/zenodo.7395871).
This DOI represents all versions, and will always resolve to the latest one.

<!--To cite the last released version, please check
https://zenodo.org/account/settings/github/repository/aumpierre-unb/Game-of-Life-for-GNU-Octave.-->

---

## The Conway's Game of Life

Conway's Game of Life computes the configuration of the next generation of a grid of square cells, each of which is in either live or dead. Each cell will be live or dead in the next generation according to two simple rules:

- A live cells with exactly 2 or 3 live neighbors stays a live cell, otherwise it dies.
- A dead cell with 3 live neighbors becomes alive, otherwise is stays dead.

As simple as these rules may be, they can result in surprising complexity from relatively simple initial configurations.

## The `game-of-life` Package

`game-of-life` provides the following functions:

- `genzero`
- `conway`
- `evolution`

### `genzero`

`genzero` generates a matrix of ones and zeros of m lines and n columns.
By default, elements of the matrix are computed individually with 50 % probability of assuming the value one.
If probability for ones if given, the values of all elements are computed with this parameter.
If the grid pattern is assigned by strings 'block', 'glider', 'blinker', 'toad' or 'beacon' matrix is generated with the pattern at its superior left corner.

**Syntax:**

```dotnetcli
[A]=genzero(m,n[,P=.50])
```

**Examples:**

Generate a spaceship glider pattern in a 9 lines and 7 columns matrix.

```dotnetcli
A=genzero(9,7,'glider')
```

Generate a 27 by 28 cells grid where individual cells have a 26:74 chance to be live or dead.

```dotnetcli
A=genzero(27,28,.26)
```

### `conway`

`conway` computes the configuration of the next generation of a grid of square cells, each of which is in either live or dead.

Each cell will be live or dead in the next generation according to the rules of the Conway's game of life.

By default, the grid is assumed to be finite. If *cyc*=*true* is given, then grid is assumed to be infinite and periodic.

By default, no plot is shown. If *fig*=*true* is given, then a figure with two subplots is shown, the first subplot being the configuration of the original grid and the second subplot being the configuration of the next generations.

By default, the colormap pattern of plots is *gray*. Please address to the colormap help documentation for different patterns.

**Syntax:**

```dotnetcli
[Re]=f2Re(f,[eps[,fig=false]])
```

**Examples:**

Compute the next generation of a spaceship glider pattern in a 10 by 10 cells grid and visualize the grids on plots.

```dotnetcli
A=genzero(10,10,'glider');...
B=conway(A,true,true)
```

### `evolution`

`evolution` simulates the evolution of a grid of square cells according to the rules of the Conway's game of life.

The simulations will be interrupted if the grid comes to a stable configuration (including death of all cells). By default, 100 generations will be computed.

By default, the grid is assumed to be finite. If *cyc*=*true* is given, then grid is assumed to be infinite and periodic.

By default, a figure with two subplots is shown, the first subplot being the configuration of the original grid and the second subplot being the configuration of the next generations. If *fig*=*false* is given, then no plot is shown.

By default, the colormap pattern of plots is *gray*. Please address to the colormap help documentation for different patterns.

By default, the next generation configuration is shown after 30 milliseconds.

**Syntax:**

```dotnetcli
[B]=evolution(A,N=100,cyc=false,fig=true,map='gray',s=30)
```

**Examples:**

Observe the evolution of the first 200 generations of a grid originally set as a spaceship glider pattern in a periodic 13 by 11 cells grid.

```dotnetcli
A=genzero(13,11,'glider');...
B=evolution(A,200,true)
```

Generate a periodic 27 by 28 cells grid where individual cells have a 26:74 chance to be live or dead and observe its evolution.

```dotnetcli
A=genzero(27,28,.26);...
B=evolution(A,inf,true)
```

Copyright &copy; 2022 2023 Alexandre Umpierre

email: aumpierre@gmail.com


NEWS
TODO
====

game-of-life 0.1.2 (2022-12-10)
======================================
* Minor changes on documentation

game-of-life 0.1.1 (2022-12-10)
======================================
* New patterns included
* Input and output plots combined as subplots
* Minor changes on functions
* Minor changes on documentation

game-of-life 0.1.0 (2022-12-08)
======================================
* First release
* Package under construction


src/figs.m
# Copyright (C) 2022 Alexandre Umpierre
#
# This file is part of game-of-life toolbox for GNU Octave.
# game-of-life toolbox for GNU Octave is free software:
# you can redistribute it and/or modify it under the terms
# of the GNU General Public License (GPL) version 3
# as published by the Free Software Foundation.
#
# game-of-life toolbox for GNU Octave is distributed in the hope
# that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the
# GNU General Public License along with this program
# (license GNU GPLv3.txt).
# It is also available at https://www.gnu.org/licenses/.

function figs(A,map)
    # Syntax:
    # figs(A,map)
    #
    # figs produces two plots
    #  the first being
    #  the configuration of the original grid and
    #  the second,
    #  the configuration of the next generation.
    # figs is an internal function of
    #  the game-of-life toolbox for GNU Octave.
    P=size(A,2);
    Q=size(A,1);
    figure('color','#A3A3A3');
    colormap(map);
    set(gca,'xlim',[.5 P+.5],...
    'ylim',[.5 Q+.5],...
    'xgrid','on',...
    'ygrid','on',...
    'gridcolor',[1 1 1],...
    'xtick',[1:P],...
    'xtick',[1:Q],...
    'box','on');
    image(A*100);
    axis('nolabel','equal');
endfunction



src/concat.m
# Copyright (C) 2022 2023 Alexandre Umpierre
#
# This file is part of game-of-life toolbox for GNU Octave.
# game-of-life toolbox for GNU Octave is free software:
# you can redistribute it and/or modify it under the terms
# of the GNU General Public License (GPL) version 3
# as published by the Free Software Foundation.
#
# game-of-life toolbox for GNU Octave is distributed in the hope
# that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the
# GNU General Public License along with this program
# (license GNU GPLv3.txt).
# It is also available at https://www.gnu.org/licenses/.

function A=concat(M,N)
    # Syntax:
    # [A]=concat(M,N)
    #
    # concat produces the
    #  horiziontal and vertical concatenation
    #  of matrices M and N.
    # concat is an internal function of
    #  the game-of-life toolbox for GNU Octave.
    #
    # Examples:
    # # Concatenate
    # # a 2 by 3 matrix of ones and
    # # a 4 by 5 matrix of random numbers.
    # M=ones(2,3),...
    # N=rand(4,5),...
    # A=concat(M,N)
    P=zeros(size(M,1),size(N,2));
    Q=zeros(size(N,1),size(M,2));
    A=[M P;Q N];
end


src/conway.m
# Copyright (C) 2022 2023 Alexandre Umpierre
#
# This file is part of game-of-life toolbox for GNU Octave.
# game-of-life toolbox for GNU Octave is free software:
# you can redistribute it and/or modify it under the terms
# of the GNU General Public License (GPL) version 3
# as published by the Free Software Foundation.
#
# game-of-life toolbox for GNU Octave is distributed in the hope
# that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the
# GNU General Public License along with this program
# (license GNU GPLv3.txt).
# It is also available at https://www.gnu.org/licenses/.

function B=conway(A,cyc=false,fig=false,map='gray')
    # Syntax:
    # [B]=conway(A[,cyc=false[,fig=false[,map='gray']]])
    #
    # conway computes
    #  the configuration of
    #  the next generation of
    #  a grid of square cells,
    #  each of which is in either live or dead.
    # Each cell will be live or dead
    #  in the next generation according to
    #  the rules of the Conway's game of life:
    # 1) A live cells with exactly 2 or 3 live neighbors
    #  stays a live cell, otherwise it dies.
    # 2) A dead cell with 3 live neighbors
    #  becomes alive, otherwise is stays dead.
    # By default, the grid is assumed to be finite.
    #  If cyc=true is given, then
    #  grid is assumed to be infinite and periodic.
    # By default, no plot is shown.
    #  If fig=true is given, then
    #  a figure with two subplots is shown,
    #  the first subplot being
    #  the configuration of the original grid and
    #  the second subplot being
    #  the configuration of the next generations.
    # By default, the colormap pattern of plots
    #  is gray. Please address to
    #  the colormap help documentation for
    #  different patterns.
    # conway is a main function of
    #  the game-of-life toolbox for GNU Octave.
    #
    # Examples:
    # # Compute the next generation
    # # of a spaceship glider pattern
    # # in a 10 by 10 cells grid and
    # # visualize the grids on plots.
    # A=genzero(10,10,'glider');...
    # B=conway(A,true,true)
    #
    # See also: genzero, evolution.
    if cyc
        C=[A(:,end),A,A(:,1)];
        D=[C(end,:);C;C(1,:)];
    else
        D=concat(0,A);
        D=concat(D,0);
    end
    B=zeros(size(D));
    for i=2:size(D,1)-1
        for j=2:size(D,2)-1
            viz=0;
            for m=-1:1
                for n=-1:1
                    if (m==0 && n==0)
                        continue;
                    end
                    if D(i+m,j+n)==1
                        viz=viz+1;
                    end
                    if viz==4
                        break;
                    end
                end
            end
            if D(i,j)==1 && viz==2 || viz==3
                B(i,j)=1;
            end
            if D(i,j)==0 && viz==3
                B(i,j)=1;
            end
        end
    end
    B=B(2:end-1,2:end-1);
    if fig
        P=size(A,2);
        Q=size(A,1);
        figure('color','#A3A3A3');subplot(1,2,1);
        colormap(map);
        set(gca,'xlim',[.5 P+.5],...
        'ylim',[.5 Q+.5],...
        'xgrid','on',...
        'ygrid','on',...
        'gridcolor',[.73 .73 .73],...
        'xtick',[1:P],...
        'xtick',[1:Q],...
        'box','on');
        image(A*100);
        axis('nolabel','equal');
        subplot(1,2,2);
        colormap(map);
        set(gca,'xlim',[.5 P+.5],...
        'ylim',[.5 Q+.5],...
        'xgrid','on',...
        'ygrid','on',...
        'gridcolor',[1 1 1],...
        'xtick',[1:P],...
        'xtick',[1:Q],...
        'box','on');
        image(B*100);
        axis('nolabel','equal');
    end
end



src/genzero.m
# Copyright (C) 2022 2023 Alexandre Umpierre
#
# This file is part of game-of-life toolbox for GNU Octave.
# game-of-life toolbox for GNU Octave is free software:
# you can redistribute it and/or modify it under the terms
# of the GNU General Public License (GPL) version 3
# as published by the Free Software Foundation.
#
# game-of-life toolbox for GNU Octave is distributed in the hope
# that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the
# GNU General Public License along with this program
# (license GNU GPLv3.txt).
# It is also available at https://www.gnu.org/licenses/.

function A=genzero(m,n,P=.50)
    # Syntax:
    # [A]=genzero(m,n[,P=.5])
    #
    # genzero generates a matrix of
    #  ones and zeros of
    #  m lines and n columns.
    # By default, elements of the matrix
    #  are computed individually with
    #  50 percent probability of
    #  assuming the value one.
    # If probability for ones if given,
    #  the values of all elements are
    #  computed accordingly.
    # If the grid pattern is assigned by strings
    #  'block', 'glider', 'blinker', 'toad' or 'beacon'
    #  matrix is generated with the pattern at
    #  its superior left corner.
    # genzero is a main function of
    #  the game-of-life toolbox for GNU Octave.
    #
    # Examples:
    # # Generate a spaceship glider pattern
    # # in a 9 lines and 7 columns matrix.
    # A=genzero(9,7,'glider')
    #
    # # Generate a 27 by 28 cells grid
    # # where individual cells have a 26:74 chance
    # # to be live or dead.
    # A=genzero(27,28,.26)
    #
    # See also: conway, evolution.
    if isreal(P) && isscalar(P) && P>0
        A=(rand(m,n)<P);
    elseif strcmpi(P,'block')
        A=[0 0 0;
           0 1 1;
           0 1 1];
        if m>size(A,1)
            A(m,end)=0;
        end
        if n>size(A,2)
            A(end,n)=0;
        end
    elseif strcmpi(P,'glider')
        A=[0 0 0 0;
           0 0 1 0;
           0 0 0 1;
           0 1 1 1];
        if m>size(A,1)
            A(m,end)=0;
        end
        if n>size(A,2)
            A(end,n)=0;
        end
    elseif strcmpi(P,'blinker')
        A=[0 0 0 0;
           0 0 1 0;
           0 0 1 0;
           0 0 1 0];
        if m>size(A,1)
            A(m,end)=0;
        end
        if n>size(A,2)
            A(end,n)=0;
        end
    elseif strcmpi(P,'toad')
        A=[0 0 0 0 0;
           0 0 0 0 0
           0 0 1 1 1;
           0 1 1 1 0;
           0 0 0 0 0];
        if m>size(A,1)
            A(m,end)=0;
        end
        if n>size(A,2)
            A(end,n)=0;
        end
    elseif strcmpi(P,'beacon')
        A=[0 0 0 0 0;
           0 1 1 0 0;
           0 1 1 0 0;
           0 0 0 1 1;
           0 0 0 1 1];
        if m>size(A,1)
            A(m,end)=0;
        end
        if n>size(A,2)
            A(end,n)=0;
        end
    end
end



src/evolution.m
# Copyright (C) 2022 2023 Alexandre Umpierre
#
# This file is part of game-of-life toolbox for GNU Octave.
# game-of-life toolbox for GNU Octave is free software:
# you can redistribute it and/or modify it under the terms
# of the GNU General Public License (GPL) version 3
# as published by the Free Software Foundation.
#
# game-of-life toolbox for GNU Octave is distributed in the hope
# that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
#
# You should have received a copy of the
# GNU General Public License along with this program
# (license GNU GPLv3.txt).
# It is also available at https://www.gnu.org/licenses/.

function [B,n]=evolution(A,N=100,cyc=false,fig=true,map='gray',s=30)
    # Syntax:
    # [B]=evolution(A[,N=100[,cyc=false[,fig=true[,map='gray'[,s=30]]]]])
    #
    # evolution simulates the evolution of
    #  a grid of square cells
    #  according to the rules of
    #  the Conway's game of life.
    # The simulations will be interrupted
    #  if the grid comes to
    #  a stable configuration (including death of all cells).
    # By default, N=100 generations will be computed.
    # By default, the grid is assumed to be finite.
    #  If cyc=true is given, then
    #  grid is assumed to be infinite and periodic.
    # By default, a figure with two subplots is shown,
    #  the first subplot being
    #  the configuration of the original grid and
    #  the second subplot being
    #  the configuration of the next generations.
    #  If fig=false is given, then no plot is shown.
    # By default, the colormap pattern of plots
    #  is gray. Please address to
    #  the colormap help documentation for
    #  different patterns.
    # By default, the next generation configuration
    #  is shown after s=30 milliseconds.
    # evolution is a main function of
    #  the game-of-life toolbox for GNU Octave.
    #
    # Examples:
    # # Observe the evolution of
    # # the first 200 generations of
    # # a grid originally set as
    # # a spaceship glider pattern in
    # # a periodic 13 by 11 cells grid.
    # A=genzero(13,11,'glider');...
    # B=evolution(A,200,true)
    #
    # # Generate a periodic 27 by 28 cells grid
    # # where individual cells have a 26:74 chance
    # # to be live or dead and
    # # observe its evolution.
    # A=genzero(27,28,.26);...
    # B=evolution(A,inf,true)
    #
    # See also: conway, genzero.
    n=0;
    C={};
    start=true;
    k=0;
    while n<N
        if n<10
            C=[C A];
            m=n;
        else
            for i=1:10-1
                C{i}=C{i+1};
            end
            C{10}=A;
            m=10;
        end
        B=conway(A,cyc,:,map);
        n=n+1;
        for i=1:m
            if all(all(B==C{i}))
                printf('Stability after %d itarations with period %d.\n',n-1,m-i+1);
                image(B*100);
                return;
            end
        end
        if n==N
            printf('Not stable after %d itarations.\n',n);
        end
        if start && fig
            P=size(A,2);
            Q=size(A,1);
            figure('color','#A3A3A3');
            subplot(1,2,1);
            colormap(map);
            set(gca,'xlim',[.5 P+.5],...
            'ylim',[.5 Q+.5],...
            'xgrid','on',...
            'ygrid','on',...
            'gridcolor',[.73 .73 .73],...
            'xtick',[1:P],...
            'xtick',[1:Q],...
            'box','on');
            image(A*100);
            axis('nolabel','equal','tic[]');
            start=false;
            subplot(1,2,2);
            colormap(map);
            set(gca,'xlim',[.5 P+.5],...
            'ylim',[.5 Q+.5],...
            'xgrid','on',...
            'ygrid','on',...
            'gridcolor',[1 1 1],...
            'xtick',[1:P],...
            'xtick',[1:Q],...
            'box','on');
        end
        if fig
            image(B*100);
            axis('nolabel','equal','tic[]');
            hold on;
            pause(s/1000);
        end
        if false
            k=k+1;
            print(sprintf('animation%04d.gif',k));
        end
        A=B;
    end
endfunction

